tools/shell/embedded_shell.cpp

```diff
@@ -504,7 +508,24 @@ std::vector<std::unique_ptr<QueryResult>> EmbeddedShell::processInput(std::strin
         ss.clear();
         ss.str(unicodeInput);
         while (getline(ss, query, ';')) {
-            queryResults.push_back(conn->query(query));
+            auto queryResult = conn->query(query);
+            auto isSuccess = queryResult->isSuccess();
+            queryResults.push_back(std::move(queryResult));
+            if (isSuccess && !database->getConfig().readOnly) {
+                auto clientContext = conn->getClientContext();
```

Comment:
This looks fine to me. @ray6080 can u take a second look?

Reply:
hmm I'm concerned about this approach because it introduces an additional binding step to every write query, which I don't think the overhead is insignificant. Need to think a bit on what's an alternative way.

Reply:
The additional binding only occurs in the shell and it should only take a few milliseconds. I don't think CLI users care about this little time.

Reply:
But I think we can limit this to only credential related queries? Instead of check db being read only, I think it's better if we can check from queryResult the statement is credential related, and this should be doable.